File: 39-combination-sum/combination-sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const ans: number[][] = [];

    function findCombination(index: number, arr: number[], sum: number) {
        // Base case: if the sum equals the target, add the combination to the result
        if (sum === target) {
            ans.push([...arr]);
            return;
        }

        // Base case: if the sum exceeds the target or index goes out of bounds, stop
        if (sum > target || index >= candidates.length) {
            return;
        }

        // Include the current candidate (can be used multiple times)
        arr.push(candidates[index]);
        sum += candidates[index];
        findCombination(index, arr, sum );
        sum -= candidates[index]; // Recurse with the same index
        arr.pop();

        // Exclude the current candidate and move to the next
        findCombination(index + 1, arr, sum);
    }

    findCombination(0, [], 0);
    return ans;
}

File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> curList = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        generate(candidates ,  target ,  0 , 0);
        return ans;
    }
    private void generate(int[] candidates , int target , int curSum ,int index){
        if(curSum == target){
            ans.add(new ArrayList(curList));
            return;
        }
        if(curSum > target || index == candidates.length) return;
        for(int i = index; i < candidates.length; i ++) {
            curList.add(candidates[i]);
            generate(candidates , target , curSum + candidates[i] , i);
            curList.remove(curList.size() - 1);
        }
    }
}